### PR TITLE
Add a WPForms contact form during one-click theme setup

### PR DESCRIPTION
**Pull request by nickcernis**
Fixes https://github.com/studiopress/genesis-sample/issues/229.

When testing the suggested code in https://github.com/studiopress/genesis-sample/commit/e0e38e5a790d26958fab92852f0aacebec313972 I found two issues:

1. **The created form is missing `id` and `submit_text` keys**. This results in a blank submit button and PHP warnings on the contact page. I've changed the way the form is created to overcome this. (The `update` method allows us to pass a full config instead of depending on the 'contact' template, which has missing data intended to be completed when someone publishes the form as the final creation step in their browser.)
2. **Multiple forms are created during a single onboarding process.** I was seeing as many as eight forms created for each run of one-click theme setup, likely because Genesis itself calls `genesis_get_config( 'onboarding' )` multiple times, so any side-effects linked to the `onboarding.php` file (like form creation) will happen multiple times. This PR overcomes that by storing the form ID as a `genesis_onboarding_wpforms_id` option when it's first created, and only adding new forms if there's no stored ID. This also allows us to re-use the StudioPress-created form for other themes, or when running one-click theme setup again for the same theme.

All of this adds more code/logic than I'd hoped, but it creates a better user experience overall and should be something we can carry to other themes if we wish.

### CODE REVIEW

File: config/wpforms.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ */
+return array(
+	'field_id' => '3',
+	'fields'   => array(
+		'0' => array(
+			'id'       => '0',
+			'type'     => 'name',
+			'label'    => esc_html__( 'Name', 'genesis-sample' ),
+			'required' => '1',
+			'size'     => 'medium',
```
Comment (dreamwhisper) [inline config/wpforms.php]:
```suggestion
			'size'     => 'large',
```

**Comment by nathanrice** [comment]
Yeah, that is a lot more code than I really wanted to add to a child theme ... much less every single child theme we build from now on.

What if we could resolve the issues you found externally?

1. Perhaps we could approach WPF and ask them to add a template where the form is fully complete?
2. This sounds like a flaw in our code in Genesis. Now is a perfect time to fix this, IMO. The onboarding config shouldn't need to be loaded multiple times. If it's happening in a single pageload, we should be able to cache the onboarding config array to use across multiple functions. If it's being called for each AJAX request, that's probably a lot harder.

File: config/wpforms.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		),
+		'1' => array(
+			'id'       => '1',
+			'type'     => 'email',
+			'label'    => esc_html__( 'Email', 'genesis-sample' ),
+			'required' => '1',
+			'size'     => 'medium',
```
Comment (dreamwhisper) [inline config/wpforms.php]:
```suggestion
			'size'     => 'large',
```

File: config/wpforms.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * WPForms settings to use when creating a form during one-click theme setup.
+ */
+return array(
+	'field_id' => '3',
+	'fields'   => array(
+		'0' => array(
+			'id'       => '0',
+			'type'     => 'name',
+			'label'    => esc_html__( 'Name', 'genesis-sample' ),
```
Comment (dreamwhisper) [inline config/wpforms.php]:
```suggestion
			'label'    => esc_html__( 'Name', 'genesis-sample' ),
			'format'   => 'simple',
```

File: config/import/content/contact.php [hunk 4]
```diff
@@ -12,8 +13,17 @@
...
+$genesis_sample_wpforms_id = studiopress_get_wpforms_id();
+
+if ( $genesis_sample_wpforms_id ) {
+	return "<!-- wp:wpforms/form-selector {\"formId\":\"{$genesis_sample_wpforms_id}\"} /-->";
+}
+
 return <<<CONTENT
 <!-- wp:paragraph -->
-<p>Add a contact form to this page with the pre-installed WPForms Lite plugin (Third Party). Learn <a href="https://my.studiopress.com/documentation/wpforms/plugin-usage/create-contact-forms-with-wpforms/" target="_blank" rel="noreferrer noopener" aria-label=" (opens in a new tab)">how to create a form using WPForms</a>.</p>
+<p>Add a contact form to this page with the WPForms Lite plugin (Third Party). Learn <a href="https://my.studiopress.com/documentation/wpforms/plugin-usage/create-contact-forms-with-wpforms/" target="_blank" rel="noreferrer noopener" aria-label=" (opens in a new tab)">how to create a form using WPForms</a>.</p>
```
Comment (dreamwhisper) [inline config/import/content/contact.php]: This content doesn't show on the created page.

**Comment by dreamwhisper** [comment]
@nickcernis Nice Job! 

I added a couple of suggestions to match the demo form.

Initially, I thought we intended the content to still show on the content page, but, that's intended. We lose the link to documentation, but I think that's ok since the first form is fully created like the demo with the other changes I suggested.

**Comment by nickcernis** [comment]
@dreamwhisper Thanks for testing and for your suggestions — I've applied those.

@nathanrice It would be great if WPForms could create a contact form during plugin install and save its ID somewhere for us to retrieve and output in the block. That would likely save us a lot of maintenance down the line.

I also agree about fixing those issues in Genesis (perhaps we could have a hook that runs after each phase, such as plugin install). In that case, though, we might still want to store the form ID to prevent forms being created multiple times when switching between StudioPress themes (assuming we're creating the form and not using one inserted by default by WPForms).

**Comment by nickcernis** [comment]
Based on forthcoming updates to WPForms and Genesis, we will need to:

1. Switch this logic:

https://github.com/studiopress/genesis-sample/blob/967bdf3da9dad3dc118daf6fa3c5878678a012b4/lib/wpforms.php#L59-L80

…to look like this (untested, will require WPForms Lite 1.5.2):

```php
$form_id = wpforms()->form->add(
	esc_html__( 'Simple Contact Form', 'genesis-sample' ),
	array(),
	array(
		'template' => 'contact',
		'builder' => false,
	)
);

return $form_id;
```

2. Remove the `wpforms.php` config file.

3. Update the form import to run inside the upcoming `genesis_onboarding_before_import_content` hook.

4. Update theme demos to match the styling (field size/type) of the default WPForms contact template.

We could optionally also remove the logic to prevent duplicate form creation, although that introduces the possibility of people importing a form during one-click theme setup for one StudioPress theme, customising the form with new notifications and other edits, then switching to another theme and having their customised form replaced by a fresh one with the same name without realising it.

One small issue I can foresee: if someone already has WPForms installed before they start onboarding but does not have WPForms 1.5.2+, the new `'builder' => false` flag proposed for 1.5.2 will have no effect and the form we create will be broken. (Onboarding doesn't update plugins if they're already installed; it just activates them.) We can either check the WPForms Pro/Lite version numbers in addition to the check for the `wpform()` function, or just accept that this could occur in a limited number of cases until existing users have updated (my preference).

**Comment by nathanrice** [comment]
Is there any way to avoid the addition of `/lib/wpforms.php` file?

**Comment by nickcernis** [comment]
@nathanrice We can do that. It would just mean we'll be creating a new form each time onboarding is run. Perhaps we do that for now and see if we get feedback about it?

We are already recreating imported pages and menu items during multiple onboarding runs. I haven't seen complaints about this so far, but it's possible Genesis should be handling de-duplication instead of child themes anyway.

**Comment by dreamwhisper** [comment]
I’d prefer not to continually add additional forms if possible. Onboarding content is a little different in that someone can activate a different theme and get a new homepage or other content. But, for forms, it will be the same each time regardless of the theme.

**Comment by jb510** [comment]
Im admittedly split on the content import issue.  Different home pages, different internal pages after switch themes ought to be import.  

However, as I think about different themes they might have different bundled forms.  This is conceived as a contact form, but there is no reason this couldn’t be leveraged for a “reservation request form”, or a “newsletter signup form”, etc...

The only dedupe key I can think of is “form title” and that’s admittedly not ideal.  Could onboarding detect this and ask the user what to do? Do you wish to “importing Foo. Foo already exists, do you wish to replace Foo, skip importing Foo, or keep both?”

**Comment by nickcernis** [comment]
I made some additional commits here to:

- Remove the config file that's no longer needed as of WPForms Lite 1.5.2.
- Refactor the form creation function and remove the ID accessor function (we can use `get_option( 'genesis_onboarding_wpforms_id' )` in place of that).

This reduces the code we had by quite a bit, but still gives us the benefit of not recreating forms that already exist:

- A form will be created during onboarding if one has not already been created by a StudioPress theme, or if one was created but deleted.
- The form's ID is stored in `genesis_onboarding_wpforms_id`.

This is ready for retesting and potential inclusion in Genesis Sample 2.10. You will need to delete WPForms Lite if you have an earlier version than 1.5.2 installed. (Onboarding will not update plugins during the setup process, and the creation of a working WPForms form from their contact form template requires 1.5.2+.)

> The only dedupe key I can think of is “form title” and that’s admittedly not ideal.

I've stored the created form ID in the `genesis_onboarding_wpforms_id` option as a way to check if the process has run before.

> Could onboarding detect this and ask the user what to do? Do you wish to “importing Foo. Foo already exists, do you wish to replace Foo, skip importing Foo, or keep both?”

We're still thinking about ways to potentially present users with choices during onboarding, but that would likely happen in Genesis and not the child theme if we go down that path.

@nathanrice Interestingly, if I replace `studiopress_create_wpforms_form();` in `config/import/content/contact.php` with `add_action( 'genesis_onboarding_before_import_content', 'studiopress_create_wpforms_form' );`, the form is not created if it does not exist already.

I think what's happening in that case is that:

1. Genesis reads the `onboarding.php` config file.
2. When that file is read, `contact.php` is required, returning the default text instead of the WPForms block, because `get_option( 'genesis_onboarding_wpforms_id' )` is currently empty.
3. `genesis_onboarding_before_import_content` fires, creating the contact form and storing the ID in the `genesis_onboarding_wpforms_id` option.
4. Genesis imports the contact page content, which hasn't been updated to reflect that there's now a form, because the config was read before `genesis_onboarding_before_import_content` happened.

By calling the `studiopress_create_wpforms_form()` function directly instead, it forces the form to be created before the contact page content is stored.

I mention this in case it comes up during or after the Genesis 2.10 beta. It raises the question of how theme devs are supposed to pass data from functions that run during `genesis_onboarding_before_import_content` to imported content (if they wish to at all). It may be that we need a hook to run earlier before any onboarding config is read.

**Comment by nathanrice** [comment]
@nickcernis your suspicion is correct. The form isn't created until after its existence is checked for.

However, there's another problem with what exists currently in this PR, and it makes my question about removing `/lib/wpforms.php` look a little silly now.

By placing the form creation code in **any portion of the code** that gets required when reading a config file, it gets immediately executed. Which means that whether or not the person actually decides to run the onboarding tool, a form will be created. This isn't cool, and I should have recognized it earlier.

To see this in action, delete all forms, activate Sample, but bypass onboarding. Unless I'm reading the code wrong, you'll see that the theme created a form, even though you skipped onboarding.

This has several implications beyond just what is being discussed here (for instance, config files can't contain expensive processes, at least configs that are checked/loaded on every pageload), but I'll limit the discussion to this particular problem on this issue.

My suggestion would be to do this:
Move the `add_action()` call to `lib/wpforms.php`, and do all the form creation stuff there (as you are now). Perhaps we can centralize this somehow in the future, I'm not sure. But for now, it's necessary.

**Comment by nickcernis** [comment]
> Move the add_action() call to lib/wpforms.php, and do all the form creation stuff there (as you are now).

@nathanrice Do you mean do `add_action( 'genesis_onboarding_before_import_content', 'studiopress_create_wpforms_form' );` in wpforms.php instead of doing `studiopress_create_wpforms_form();` in contact.php?

That solves the issue of the form being created even if onboarding isn't running, but it doesn't fix the issue of forms not being created in time for them to be used in the contact form content.

One way to solve both issues might be to:

1. Create the form with the placeholder content by default.
2. Hook to genesis_onboarding_**after**_import_content with a function that (a) creates the form (b) replaces the contact form placeholder content with the WPForms block (the contact page ID is exposed to `genesis_onboarding_after_import_content` via the passed `$imported_post_ids` param, so would be easy to fetch and update.)

Does that sound like a better approach here, or is there something else we should try first?

**Comment by nathanrice** [comment]
>That solves the issue of the form being created even if onboarding isn't running, but I don't think it fixes the issue of forms not being created in time for them to be used in the contact form content

Sorry, I assumed `lib/wpforms.php` was being `require`d in `functions.php`. That would ensure that the code in `lib/wpforms.php` runs (and the action callback is attached) before the onboarding code runs, correct?

**Comment by nickcernis** [comment]
> That would ensure that the code in lib/wpforms.php runs (and the action callback is attached) before the onboarding code runs, correct?

Unfortunately it does not seem to, no. The action callback is attached earlier, but the contact form import code still runs after the config file is read, because the `genesis_onboarding_after_import_content` hook fires after we read the onboarding config.

**Comment by nathanrice** [comment]
>The action callback is attached earlier, but the contact form import code still runs after the config file is read, because the genesis_onboarding_after_import_content hook fires after we read the onboarding config.

Right. OK. Let me think a bit on this and see if there's a simpler approach. Might have to make some changes in Genesis core. We have til Monday (beta2), so that gives today and tomorrow to experiment a bit.

**Comment by nathanrice** [comment]
I'm writing this out to help myself follow the problem better ...

We need to be able to create a form.

This form creation must happen ONLY if the user has elected to run the onboarding tool, otherwise the form will be created needlessly.

But this form creation must happen early enough for the config file (when first read) to know of the form's existence.

I honestly don't see how this is possible. Because the code in the config isn't just content, it's executable code. Once it executes once (on first read) it doesn't need to execute again. So our conditionals won't work.

It might be that we have to do exactly what you suggested above, run a procedure after onboarding that creates a form and replaces the placeholder with the form code. Let's be sure to discuss this in our meeting about updatable child themes.

**Comment by nickcernis** [comment]
@nathanrice @dreamwhisper I've updated this to swap out the contact page placeholder content with a form block during `genesis_onboarding_after_import_content`, as we discussed.

I believe this solves all the issues we've encountered with this so far, but would appreciate a further review when you have time. You should find that:

- A contact form is created only if onboarding is run and no contact form has been created by an SP theme before. (I've been using `wp option delete genesis_onboarding_wpforms_id` to clear the stored ID during testing.)

- The placeholder content will be replaced by a working forms block by default.

- The placeholder content will be shown if WPForms Lite is removed from the `onboarding.php` config file and all WPForms plugins (Lite/Pro) are deactivated.

**Comment by dreamwhisper** [comment]
@nickcernis Looks good!

@nathanrice I'll wait to merge until you've taken a look.

**Pull request merged**